ipc/common/messenger.go

```diff
@@ -0,0 +1,183 @@
+	if err != nil {
+		return err
+	}
+
+	err = messenger.sendMessageLengthAndKind(len(dataBytes), message.GetKind())
+	if err != nil {
+		return err
+	}
+
+	_, err = messenger.writer.Write(dataBytes)
+	if err != nil {
+		return err
+	}
+
+	messenger.Logger.Trace(fmt.Sprintf("[%s][#%d]: SENT message", messenger.Name, message.GetNonce()), "size", len(dataBytes), "msg", message)
+	return err
+}
+
+func (messenger *Messenger) sendMessageLengthAndKind(length int, kind MessageKind) error {
+	buffer := make([]byte, 8)
+	binary.LittleEndian.PutUint32(buffer[0:4], uint32(length))
+	binary.LittleEndian.PutUint32(buffer[4:8], uint32(kind))
+	_, err := messenger.writer.Write(buffer)
+	return err
+}
+
+// Receive receives a message, reads it from the pipe
+func (messenger *Messenger) Receive(timeout int) (MessageHandler, error) {
+	messenger.Logger.Trace(fmt.Sprintf("[%s]: Receive message...", messenger.Name))
+
+	if timeout != 0 {
+		messenger.setReceiveDeadline(timeout)
+		defer messenger.resetReceiveDeadline()
+	}
+
+	length, kind, err := messenger.receiveMessageLengthAndKind()
+	if err != nil {
+		return nil, err
+	}
+
+	message := CreateMessage(kind)
+
+	// Now read the body
+	buffer := make([]byte, length)
+	_, err = io.ReadFull(messenger.reader, buffer)
+	if err != nil {
+		return nil, err
+	}
+
+	err = messenger.unmarshal(buffer, message)
+	if err != nil {
+		return nil, err
+	}
+
+	messenger.Logger.Trace(fmt.Sprintf("[%s][#%d]: RECEIVED message", messenger.Name, message.GetNonce()), "size", length, "msg", message)
+	messageNonce := message.GetNonce()
+	if messageNonce != messenger.Nonce+1 {
+		return nil, ErrInvalidMessageNonce
+	}
+
+	messenger.Nonce = messageNonce
+	return message, nil
+}
+
+func (messenger *Messenger) setReceiveDeadline(timeout int) {
+	duration := time.Duration(timeout) * time.Millisecond
+	future := time.Now().Add(duration)
+	messenger.reader.SetDeadline(future)
+}
+
+func (messenger *Messenger) resetReceiveDeadline() {
+	messenger.reader.SetDeadline(time.Time{})
+}
+
+func (messenger *Messenger) receiveMessageLengthAndKind() (int, MessageKind, error) {
+	buffer := make([]byte, 8)
+	_, err := io.ReadFull(messenger.reader, buffer)
+	if err != nil {
+		return 0, FirstKind, err
+	}
+
+	length := binary.LittleEndian.Uint32(buffer[0:4])
+	kind := MessageKind(binary.LittleEndian.Uint32(buffer[4:8]))
+	return int(length), kind, nil
+}
+
+// Shutdown closes the pipes
+func (messenger *Messenger) Shutdown() {
+	messenger.Logger.Debug("%s:  Messenger::Shutdown", messenger.Name)
+
+	err := messenger.writer.Close()
+	if err != nil {
+		messenger.Logger.Error("Cannot close writer: %v", err)
+	}
+
+	err = messenger.reader.Close()
+	if err != nil {
+		messenger.Logger.Error("Cannot close reader: %v", err)
+	}
+}
+
+// EndDialogue resets the dialogue nonce
+func (messenger *Messenger) EndDialogue() {
+	messenger.Nonce = 0
+}
+
+func (messenger *Messenger) marshal(data interface{}) ([]byte, error) {
+	return marshalJSON(data)
```

Comment:
messenger should have a marshalizer and call that directly.

Reply:
(Previous comment)

Reply:
Fixed.